From c457b594aa29b6980e3887ee309d2df96a89884c Mon Sep 17 00:00:00 2001
From: Juan RP <xtraeme@voidlinux.eu>
Date: Wed, 14 Oct 2015 11:46:56 +0200
Subject: [PATCH] libxbps: fix eudev-libgudev -> libgudev fallout.

Packages to be replaced must be collected *before* checking for
broken reverse dependencies, otherwise them will always be detected
as broken.
---
 NEWS                              |  3 +++
 lib/transaction_dictionary.c      | 16 ++++++++--------
 lib/transaction_package_replace.c |  1 +
 lib/transaction_revdeps.c         |  4 ++++
 4 files changed, 16 insertions(+), 8 deletions(-)

--- lib/transaction_dictionary.c
+++ lib/transaction_dictionary.c
@@ -300,6 +300,14 @@ xbps_transaction_prepare(struct xbps_handle *xhp)
 	xbps_object_release(edges);
 
 	/*
+	 * Check for packages to be replaced.
+	 */
+	if ((rv = xbps_transaction_package_replace(xhp, pkgs)) != 0) {
+		xbps_object_release(xhp->transd);
+		xhp->transd = NULL;
+		return rv;
+	}
+	/*
 	 * If there are missing deps or revdeps bail out.
 	 */
 	xbps_transaction_revdeps(xhp, pkgs);
@@ -331,14 +339,6 @@ xbps_transaction_prepare(struct xbps_handle *xhp)
 		}
 	}
 	/*
-	 * Check for packages to be replaced.
-	 */
-	if ((rv = xbps_transaction_package_replace(xhp, pkgs)) != 0) {
-		xbps_object_release(xhp->transd);
-		xhp->transd = NULL;
-		return rv;
-	}
-	/*
 	 * Add transaction stats for total download/installed size,
 	 * number of packages to be installed, updated, configured
 	 * and removed to the transaction dictionary.
--- lib/transaction_package_replace.c
+++ lib/transaction_package_replace.c
@@ -121,6 +121,7 @@ xbps_transaction_package_replace(struct xbps_handle *xhp, xbps_array_t pkgs)
 			 */
 			xbps_dictionary_set_cstring_nocopy(instd,
 			    "transaction", "remove");
+			xbps_dictionary_set_bool(instd, "replaced", true);
 			if (!xbps_array_add_first(pkgs, instd)) {
 				xbps_object_iterator_release(iter);
 				free(pkgname);
--- lib/transaction_revdeps.c
+++ lib/transaction_revdeps.c
@@ -166,6 +166,10 @@ xbps_transaction_revdeps(struct xbps_handle *xhp, xbps_array_t pkgs)
 			 * the transaction.
 			 */
 			if (strcmp(tract, "remove") == 0) {
+				if (xbps_dictionary_get(obj, "replaced")) {
+					free(pkgname);
+					continue;
+				}
 				if (xbps_find_pkg_in_array(pkgs, pkgname, "remove")) {
 					free(pkgname);
 					continue;
-- 
2.6.1

